code/modules/chemistry/Chemistry-Dispenser.dm

```diff
@@ -271,224 +128,167 @@
-				return
-			//DEBUG(reagents)
-			var/list/reagentlist = params2list(reagents)
-			//DEBUG(list2params(reagentlist))
-
-			var/datum/reagent_group/G = new /datum/reagent_group()
-			for (var/reagent in reagentlist)
-				if (lowertext(reagent) in src.dispensable_reagents)
-					G.reagents += lowertext(reagent)
-					//Special amounts!
-					if (istext(reagentlist[reagent])) //Set a dispense amount
-						var/num = text2num(reagentlist[reagent])
-						if(!num) num = 10
-						G.reagents[lowertext(reagent)] = clamp(round(num), 1, 100)
-					else //Default to 10 if no specific amount given
-						G.reagents[lowertext(reagent)] = 10
-
-			if(G.reagents == 0)
-				doing_a_thing = 0
-				return
-
-			G.name = name
-			G.update_desc()
-			if (current_account)
-				current_account.groups += G
-			send_group_details()
-
-			doing_a_thing = 0
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if(!ui)
+			ui = new(user, src, "ChemDispenser", src.name)
+			ui.open()
+
+	ui_static_data(mob/user)
+		var/list/static_data = list()
+		var/list/dispensableReagentsTemp = list()
+		if(dispensable_reagents)
+			for(var/reagent in dispensable_reagents)
+				var/datum/reagent/current_reagent = reagents_cache[reagent]
+				dispensableReagentsTemp.Add(list(list(
+					name = current_reagent.name,
+					colorR = current_reagent.fluid_r,
+					colorG = current_reagent.fluid_g,
+					colorB = current_reagent.fluid_b,
+					state = current_reagent.reagent_state,
+					id = reagent
+				)))
+		static_data["beakerName"] = glass_name
+		static_data["dispensableReagents"] = dispensableReagentsTemp
+
+		return static_data
+
+	ui_data(mob/user)
+		var/list/data = list()
+		var/list/beakerContentsTemp = list()
+		var/list/groupListTemp = list()
+		data["idCardInserted"] = !isnull(src.user_id)
+		data["idCardName"] = !isnull(src.user_id) ? src.user_id.registered : "None"
+		data["maximumBeakerVolume"] = (!isnull(beaker) ? beaker.reagents.maximum_volume : 0)
+		data["beakerTotalVolume"] = (!isnull(beaker) ? beaker.reagents.total_volume : 0)
+		if(beaker)
+			var/datum/reagents/R = beaker:reagents
+			var/datum/color/average = R.get_average_color()
+			data["currentBeakerName"] = beaker.name
+			data["finalColor"] = average.to_rgba()
+			if(istype(R) && R.reagent_list.len>0)
+				for(var/reagent in R.reagent_list)
+					var/datum/reagent/current_reagent = R.reagent_list[reagent]
+					beakerContentsTemp.Add(list(list(
+						name = reagents_cache[reagent],
+						id = reagent,
+						colorR = current_reagent.fluid_r,
+						colorG = current_reagent.fluid_g,
+						colorB = current_reagent.fluid_b,
+						state = current_reagent.reagent_state,
+						volume = current_reagent.volume
+					)))
+		if(current_account)
+			for (var/datum/reagent_group/group in current_account.groups)
+				groupListTemp.Add(list(list(
+					name = group.name,
+					info = group.group_desc,
+					ref = ref(group)
+				)))
+		data["groupList"] = groupListTemp
```

Comment:
Probably worth considering putting these groups in `ui_static_data` and updating when relevant? Groups aren't added/removed very often, so would avoid rebuilding these lists.